docs/web3.main.rst

```diff
@@ -410,3 +410,68 @@ Each ``web3`` instance also exposes these namespaced APIs.
 .. py:attribute:: Web3.parity
 
     See :doc:`./web3.parity`
+
+
+Attaching Modules
+~~~~~~~~~~~~~~~~~
+
+Modules that inherit from the ``web3.module.Module`` class may be attached to the ``Web3`` instance either at
+instantiation or by making use of the ``attach_module()`` method.
+
+To instantiate the ``Web3`` instance with external modules:
+
+.. code-block:: python
+
+    >>> from web3 import Web3, EthereumTesterProvider
+    >>> w3 = Web3(
+    ...     EthereumTesterProvider(),
+    ...     external_modules={
+    ...         # ModuleClass objects in this example inherit from the `web3.module.Module` class
+    ...         'module1': ModuleClass1,
+    ...         'module2': (ModuleClass2, {
+    ...             'submodule1': ModuleClass3,
+    ...             'submodule2': (ModuleClass4, {
+    ...                 'submodule2a': ModuleClass5,  # submodule children may be nested further if necessary
+    ...             })
+    ...         })
+    ...     }
+    ... )
+
+    # `return_zero`, in this case, is an example attribute of the `ModuleClass1` object
+    >>> w3.module1.return_zero
+    0
+    >>> w3.module2.submodule1.return_one
+    1
+    >>> w3.module2.submodule2.submodule2a.return_two
+    2
+
+
+.. py:method:: w3.attach_module(module_name, module)
+
+    The ``attach_module()`` method can be used to attach an external module after the ``Web3`` instance has been
```

Comment:
It looks like you can only attach one top-level module here? It may be nice to be able to pass a few if that's not too difficult of a refactor. For example: 

`w3.attach_module({'eth': Eth, 'geth': Geth})`

Reply:
I think this could be useful, I was just going by the original issue raised which was to add a single module by just passing in a simple name and class. I'm open to making this more complex though if we see the benefit in it. I also don't know how many modules we will be attaching in the future and if passing in a whole dict after instantiation is really a desired thing?

I liked the simplicity of the issue raised to just passing in (`module_name`, `ModuleClass`). But yeah, I'm open to the idea of passing in a dict.

Reply:
I'm fine leaving this as attaching a single module. We can always re-think if we get an issue/feature request about it.

Reply:
Just noting here what we all discussed... it probably does makes more sense to just open this up and use a `dict` as the argument. This change was pushed in the last commit.